### DISCUSSION TITLE
High Performance Point Cloud Rendering

### DISCUSSION THREAD

**Discussion by benjaminrwilson** [bevyengine/bevy#7366] [Q&A] [answered]
Hi,

What is the most performant way to render large point clouds (100,000+ points)? I've had success by storing the points in `Mesh::new(PrimitiveTopology::PointList)`, but I haven't found a way to change the point size or to have _per point_ coloring.

Is this possible with this approach? If not, what would it require?

Thanks!

**Comment by torsteingrindvik** [comment, marked as answer]
I'm not sure about _most performant_.
But I think you'll be able to achieve this by using a storage buffer written on the CPU and copied and bound in the GPU.
If there is 1 color per point and 1 vertex per point you should be able to use the vertex index directly to read the correct color in the shader.

**Reply by superdump** [reply]
I did some hacks here for drawing many quads: https://github.com/superdump/bevy-vertex-pulling/blob/main/examples/quads/main.rs The approach is basically what @torsteingrindvik was suggesting.

**Reply by benjaminrwilson** [reply]
Awesome. Thanks @torsteingrindvik and @superdump! The advice, example, and https://github.com/bevyengine/bevy/blob/main/examples/shader/shader_instancing.rs got me what I needed.

**Comment by ottah** [comment]
If you want really performant point clouds you could write a compute rasterizer and output to a texture.

Theres some linked papers in this repo https://github.com/m-schuetz/compute_rasterizer

**Reply by benjaminrwilson** [reply]
Thanks, I'll have a look!

**Reply by geoproces** [reply]
Hi Benjamin , I am land surveyor working with points clouds ,is there any chance you can send me example of point clouds to be   imported tu bevy ?

**Reply by mosure** [reply]
see https://github.com/mosure/bevy_gaussian_splatting

**Reply by geoproces** [reply]
Thx. mousure , however I am using Lidar scarer ,not a camera , the data are  process and calculate  in hardware soft .  I just need display them in Bevy . The gaussian_splatting is very promising , but as i sett i do not have any pictures . Any ideas ?. Best Regards

**Comment by hacknus** [comment]
In case this is useful for anybody: I created a crate for a voxel plot, mainly built around the custom shading example from bevy, but with transparency support: https://github.com/hacknus/bevy_voxel_plot

Can easily handle 1000000 voxels.

**Comment by rlamarche** [comment]
Hi, I developed a point cloud plugin for Bevy : https://github.com/rlamarche/bevy_pointcloud
It's the beginning and I'm using shader instancing instead of a storage buffer.

I'm not experienced enough to tell what is the most efficient way to do it, but I'm quite satisfied by the fact that my implementations uses Bevy's frustum by computing AABB.

**Reply by hacknus** [reply]
Cool! Does it support alpha lower than 1?

**Reply by rlamarche** [reply]
Not yet. First I would like to implement high quality rendering and eye dome lightning from this paper : https://www.cg.tuwien.ac.at/research/publications/2016/SCHUETZ-2016-POT/SCHUETZ-2016-POT-thesis.pdf

For transparency, it depends what is the objective. It could be an addition of all splats, then normalization.
Or "depth peeling" that would need N passes (slow) depending on the complexity of the point cloud.

For example, if the point cloud is from LIDAR or photogrametry, we only have the surfaces. But if you're looking for something to see "inside" the point cloud, it's more complicated, and normalization would be a good alternative.

Anyway, I will give a try for all these techniques, because I'm learning shaders, and it is a very good exercise !

**Reply by hacknus** [reply]
Ah, cool! Yeah, there exist some other projects doing AABB with bevy and point clouds/voxel plots ( e.g. https://github.com/ForesightMiningSoftwareCorporation/bevy_aabb_instancing, however this is not up to date with bevy afaik). But I think AABB does not support alpha < 1 in general. My [bevy_voxel_plot](https://github.com/hacknus/bevy_voxel_plot) project supports alpha < 1 but it is less efficient and struggles with more than 1 million datapoints. So if you manage to make some progress in that direction let me know!

My objective is to render 3D scans, where I map the intensity of a pulse (reflected) to the opacity. This let's me visualize layers/reflections inside the object.
